pkg/syncer/cnsoperator/controller/cnsunregistervolume/cnsunregistervolume_controller.go

```
@@ -193,7 +198,214 @@ func (r *ReconcileCnsUnregisterVolume) Reconcile(ctx context.Context,
+			pv.Spec.PersistentVolumeReclaimPolicy = v1.PersistentVolumeReclaimRetain
+			retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				_, updateErr := k8sclient.CoreV1().PersistentVolumes().Update(context.TODO(), pv, metav1.UpdateOptions{})
+				return updateErr
+			})
+			if retryErr != nil {
+				log.Errorf("Unable to update ReclaimPolicy on PV %q", pvName)
+				return reconcile.Result{}, err
+			}
+			log.Infof("Updated ReclaimPolicy on PV %q to %q", pvName, v1.PersistentVolumeReclaimRetain)
+		}
+	} else {
+		log.Infof("CNS metadata for volume %s has missing pvName."+
+			"PV may have already been deleted. Continuing with other operations..", instance.Spec.VolumeID)
+	}
+
+	// Delete PVC.
+	if pvcName != "" && namespace != "" {
+		err = k8sclient.CoreV1().PersistentVolumeClaims(namespace).Delete(ctx,
+			pvcName, *metav1.NewDeleteOptions(0))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Errorf("Failed to delete PVC %q in namespace %q with error - %s",
+					pvcName, namespace, err.Error())
+				return reconcile.Result{}, err
+			}
+		}
+		log.Infof("Deleted PVC %q in namespace %q", pvcName, namespace)
+	} else {
+		log.Infof("CNS metadata for volume %s has missing pvcName or namespace."+
+			"PVC may have already been deleted. Continuing with other operations..", instance.Spec.VolumeID)
+	}
+
+	if pvName != "" {
+		// Delete PV.
+		// Since reclaimPolicy was set to Retain, we need to explicitly delete it.
+		err = k8sclient.CoreV1().PersistentVolumes().Delete(ctx, pvName, *metav1.NewDeleteOptions(0))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Errorf("Failed to delete PV %q", pvName)
+				return reconcile.Result{}, err
+			}
+		}
+		log.Infof("Deleted PV %q", pvName)
+	}
+
+	// Invoke CNS DeleteVolume API with deleteDisk flag set to false.
+	_, err = r.volumeManager.DeleteVolume(ctx, instance.Spec.VolumeID, false)
+	if err != nil && !cnsvsphere.IsNotFoundError(err) {
```

log the errors, silently failing would be confusing when triaging.

---

Addressed all the logging related comments in the latest commit. PTAL.